### ISSUE TITLE
merge_and_unload() raises the loss of accuracy when inference

### ISSUE THREAD

**Issue by haobozhang** [huggingface/peft#1388]
### System Info

Linux w/ peft==0.7.1

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

Here is the script:
```python
'''train and save:'''
peft_model = LlamaForCasualLM.from_pretrained(..., torch_dtype=torch.bfloat16)
peft_config = LoraConfig(
    task_type=TaskType.CAUSAL_LM, inference_mode=False, r=8, lora_alpha=32, lora_dropout=0.0, 
    target_modules=["q_proj", "k_proj", "v_proj"], bias='none',
)
peft_model.gradient_checkpointing_enable()
peft_model = get_peft_model(peft_model, peft_config)
peft_model.train()
'''...train model...'''
peft_model.eval()
peft_model.save_pretrained(save_path, save_adapter=True, save_config=True)
peft_loss = peft_model(**inputs).loss

'''load and inference:'''
load_model = PeftModel.from_pretrained(
    model=LlamaForCausalLM.from_pretrained(...), model_id=save_path, torch_dtype=torch.bfloat16)
)
load_model.eval()
load_model.merge_and_unload()
load_loss = load_model(**inputs).loss
``` 

What I get is, 
```python
peft_loss = 2.1177830696105957
load_loss = 2.121424674987793
```

I check the parameters and find $W_{load} \ne W_{peft} + BA$, which means `merge_and_unload()` induces a loss of accuracy in model parameters. 
Though the loss is affordable, I am just curious in what situation should I use `merge_and_unload()`, since it seems unnecessary to use this function when inference.

### Expected behavior

A loss of accuracy after using `merge_and_unload()`

**Comment by younesbelkada** [comment]
Hi @haobozhang 
I am afraid this is expected, although the mathematical formulation guarantees exact results, the fact that you merged the adapter weights in a bf16 model introduces some rounding errors. Can you try the same with a fp32 model?

**Comment by BenjaminBossan** [comment]
+1 to what Younes mentioned.

> Though the loss is affordable, I am just curious in what situation should I use `merge_and_unload()`, since it seems unnecessary to use this function when inference.

The main advantage is that the inference should be a little bit faster. After merging, it should have the same speed as inferencing directly with the base model, without the small overhead added by using LoRA.

I'll close the issue for now, feel free to re-open if you have further questions.
